File: Matlab/MV_rf.m
% MV_rf.m
% This script implements Mean-Variance portfolio optimization with a
% risk-free asset, deriving and plotting the Capital Market Line (CML).
%
% Developed for: Essentials of Financial Economics
% Authors: Michael Donadelli, Michele Costola, Ivan Gufler
% Date: May 8, 2025
%%

clear
clc

% Read asset returns from an Excel file
[Ret] = readtable('../Data/Returns.xlsx', VariableNamingRule='preserve');
% Read Fama-French factors from an Excel file
Factors = readtable('../Data/FF_Factors.xlsx', VariableNamingRule='preserve');

% Extract risk-free rate (assuming it's in the 5th column, starting from the 2nd row)
% Convert from annual percentage to monthly decimal
Rf = table2array(Factors(2:end, 5)) / 1200;

% Extract asset return data (assuming it starts from the 2nd column)
R = table2array(Ret(:, 2:end));
N = size(R, 2); % Number of assets

% Calculate mean returns
z = (mean(R, 1))';
% Calculate standard deviations
sig = std(R, 1);
% Calculate the covariance matrix of returns
V = cov(R);
% Calculate the inverse of the covariance matrix
V1 = inv(V);

% Calculate parameters for the efficient frontier with a risk-free asset
% H = (z - rf*1)' * inv(S) * (z - rf*1)
H = (z - ones(N, 1) * mean(Rf))' * V1 * (z - ones(N, 1) * mean(Rf));
% sqrt(H) is related to the slope of the Capital Market Line
sqrtH = sqrt(H);

% Define a range of target portfolio expected returns for plotting the CML
mu_p = linspace(mean(Rf), 0.009, 100);

% Calculate the variance of portfolios on the Capital Market Line
sig2_p = 1 / H * (mu_p - mean(Rf)).^2;
% Calculate the standard deviation (risk) of portfolios on the CML
sig_p = sqrt(sig2_p);

% Calculate the expected return and standard deviation of the Tangency Portfolio
mu_t = mean(Rf) + H / (ones(1, N) * V1 * (z - ones(N, 1) * mean(Rf)));
sig_t = sqrtH / (ones(1, N) * V1 * (z - ones(N, 1) * mean(Rf)));

% Recalculate parameters A, B, C, D for the risky asset efficient frontier (for plotting comparison)
A = z' * V1 * z;
B = z' * V1 * ones(N, 1);
C = ones(1, N) * V1 * ones(N, 1);
D = A * C - B^2;
% Define a range of target expected returns for the risky asset efficient frontier
mu_p = linspace(0.001, 0.008, 100);

% Calculate the variance of the risky asset efficient frontier
sig2_p = (1 / D) * (C * mu_p.^2 - 2 * B * mu_p + A);
sig_pp = sqrt(sig2_p);
% Set standard deviations below a certain threshold to NaN for cleaner plotting
sig_pp(mu_p < 0.005) = NaN;
x = [sig_pp; mu_p]; % Combine for potential use (though not directly used in plot)

%% Plotting the Capital Market Line and Efficient Frontier

p = figure(2);
p.WindowState = 'maximized'; % Maximize the plot window

% Plot the Capital Market Line (CML)
plot(sig_p, mean(Rf) + sqrtH .* sig_p, '-b', 'LineWidth', 1.5);
hold on
ax = gca;
ax.YAxis.Exponent = 0; % Display y-axis labels without scientific notation

% Plot the risky asset efficient frontier
plot(sig_pp, mu_p, '-k', 'LineWidth', 1.5);

% Scatter plot of the Tangency Portfolio
scatter(sig_t, mu_t, "filled")

title('Capital market line', 'FontSize', 16)
xlabel('Portfolio Risk', 'FontSize', 16)
ylabel('Portfolio Expected Return', 'FontSize', 16)
legend("Efficient frontier", "Portfolio", "Tangency port.", "Autoupdate", "off", 'FontSize', 16, Location="northwest")
xlim([0, 0.05]) % Set x-axis limits

% Add dashed lines to highlight the Tangency Portfolio's position
line([0, sig_t], [mu_t, mu_t], 'LineStyle', '--', 'Color', 'k', 'Linewidth', 1.5)
% Calculate the lower part of the CML (borrowing/lending at risk-free rate)
Ep_ = mean(Rf) - sqrtH .* sig_p;
Ep_(Ep_ < -0.002) = NaN; % Set values below a threshold to NaN for cleaner plotting
plot(sig_p, Ep_, '-b', 'LineWidth', 1.5);
line([sig_t, sig_t], [min(Ep_), mu_t], 'LineStyle', '--', 'Color', 'k', 'Linewidth', 1.5)
% text(-0.003,mu_t,"$\sigma_T$",Interpreter="latex",FontSize=12) % Example of adding text annotation

saveas(p, "MV_rf.eps", "epsc") % Save the plot as an EPS file

%% Optimal weights Calculation

% Define a range of target expected returns for which to calculate optimal weights
mup = linspace(0.001, 0.05, 10);
% Calculate optimal weights for portfolios on the CML (combination of risk-free and tangency)
wp = V1 * (z - ones(N, 1) * mean(Rf)) * (mup - mean(Rf)) / H;

% Calculate the weights for the Tangency Portfolio (risky assets only)
wT = (V1 * (z - ones(N, 1) * mean(Rf))) / (ones(N, 1)' * V1 * (z - ones(N, 1) * mean(Rf)));

% Calculate the weight allocated to the risk-free asset (1 - sum of risky asset weights)
rf = 1 - sum([wT, wp]);

% Create a table to store the calculated weights
Weights = table('Size', [6, 11], 'VariableTypes', repmat({'double'}, [1, 11]), 'VariableNames', ["wT", round(mup * 100, 2) + "%"], ...
    'RowNames', [Ret.Properties.VariableNames(2:end), "Rf"]);

% Populate the weights table with Tangency Portfolio weights and optimal CML portfolio weights (including risk-free)
Weights(:, :) = array2table(round([[wT, wp]; rf], 3));

% Write the weights table to an Excel file
writetable(Weights, "Weights_MVrf.xlsx", "FileType", "spreadsheet", "WriteVariableNames", true, ...
    "WriteRowNames", true);

% This MATLAB script contains code examples and material from the book:
% "Essentials of Financial Economics" by Michael Donadelli, Michele Costola, and Ivan Gufler.
% You can find more information and download the book at: link.springer.com/book/9783031861895

File: Python/MV_rf.py
# MV_rf.py
# This script implements Mean-Variance portfolio optimization with a
# risk-free asset, deriving and plotting the Capital Market Line (CML) in Python.
#
# Developed for: Essentials of Financial Economics
# Authors: Michael Donadelli, Michele Costola, Ivan Gufler
# Date: May 8, 2025

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
# Assuming getuncef.py is available for portfolio calculations if needed
# from getuncef import getuncef # Uncomment if getuncef is used

# --- Data Loading and Preparation ---
# Import data from Excel files
# Loading data assuming files are in the '../Data/' directory relative to this script
Ret = pd.read_excel('../Data/Returns.xlsx')
Factors = pd.read_excel('../Data/FF_Factors.xlsx')

# Extract risk-free rate (assuming it's in the 5th column, index 4, starting from the 2nd row, index 1)
# Convert from annual percentage to monthly decimal
Rf = np.array(Factors.iloc[1:, 4]) / 1200

# Extract asset return data (assuming it starts from the 2nd column, index 1)
R = np.array(Ret.iloc[:, 1:])
N = R.shape[1] # Number of assets

# Calculate mean returns (expected returns)
# Using reshape(-1, 1) to ensure it's a column vector for robust matrix operations
z = np.mean(R, axis=0).reshape(-1, 1) # Reshape to a column vector
# Calculate standard deviations (risk)
sig = np.std(R, axis=0)
# Calculate the covariance matrix of returns (rowvar=False assumes columns are variables)
V = np.cov(R, rowvar=False)
# Calculate the inverse of the covariance matrix
V1 = np.linalg.inv(V)

# --- Calculate Parameters for the Capital Market Line (CML) ---
# Calculate H, which is related to the squared Sharpe Ratio of the tangency portfolio
# H = (z - rf*1)' * inv(S) * (z - rf*1)
# Using proper matrix multiplication with z as a column vector
H = (z - np.mean(Rf)).T @ V1 @ (z - np.mean(Rf))
sqrtH = np.sqrt(H) # Related to the slope of the CML

# Define a range of target portfolio expected returns for plotting the CML
mu_p = np.linspace(np.mean(Rf), 0.009, 100)

# Calculate the variance of portfolios on the Capital Market Line
# Formula: (1/H) * (mu_p - rf)^2
sig2_p = (1 / H) * (mu_p - np.mean(Rf))**2
# Calculate the standard deviation (risk) of portfolios on the CML
sig_p = np.sqrt(sig2_p)

# Calculate the expected return and standard deviation of the Tangency Portfolio
# mu_t = rf + H / (1' * inv(S) * (z - rf*1))
# Using proper matrix operations
mu_t = np.mean(Rf) + H / (np.ones((1, N)) @ V1 @ (z - np.ones((N, 1)) * np.mean(Rf)))
# sig_t = sqrt(H) / (1' * inv(S) * (z - rf*1))
# Using proper matrix operations
sig_t = sqrtH / (np.ones((1, N)) @ V1 @ (z - np.ones((N, 1)) * np.mean(Rf)))


# --- Calculate Parameters for the Risky Asset Efficient Frontier (for comparison) ---
# Recalculate parameters A, B, C, D for the risky asset efficient frontier
# Using proper matrix operations with z as a column vector
A = z.T @ V1 @ z
B = np.ones((1, N)) @ V1 @ z
C = np.ones((1, N)) @ V1 @ np.ones((N, 1))
D = A * C - B**2

# Define a range of target expected returns for the risky asset efficient frontier
mu_p_risky = np.linspace(0.001, 0.008, 100)

# Calculate the variance of the risky asset efficient frontier
sig2_p_risky = (1 / D) * (C * mu_p_risky**2 - 2 * B * mu_p_risky + A)
sig_pp_risky = np.sqrt(sig2_p_risky)
# Set standard deviations below a certain threshold to NaN for cleaner plotting
# Using np.where for robust conditional assignment
sig_pp_risky = np.where(mu_p_risky < 0.005, np.nan, sig_pp_risky)
# x = np.vstack((sig_pp_risky, mu_p_risky)) # Combine for potential use (though not directly used in plot)


# --- Plotting the Capital Market Line and Efficient Frontier ---
fig = plt.figure(2) # Use figure 2 as in the original script
# Plot the Capital Market Line (CML)
plt.plot(sig_p.flatten(), np.mean(Rf) + sqrtH.flatten() * sig_p.flatten(), '-b', linewidth=1.5) # Ensure 1D for plotting
# Plot the risky asset efficient frontier
plt.plot(sig_pp_risky.flatten(), mu_p_risky.flatten(), '-k', linewidth=1.5) # Ensure 1D for plotting
# Scatter plot of the Tangency Portfolio
plt.scatter(sig_t.flatten(), mu_t.flatten(), marker='o', color='k') # Ensure 1D for plotting

# Configure plot appearance
plt.title('Capital market line', fontsize=16) # Title as in original
plt.xlabel('Portfolio Risk', fontsize=16) # X-label as in original
plt.ylabel('Portfolio Expected Return', fontsize=16) # Y-label as in original
# Use the exact legend entries from the original script
# Note: Some of these labels ('Portfolio', 'Stocks', 'Autoupdate', 'off') might not directly correspond
# to plotted elements in this figure, but are included to match the original legend call.
plt.legend(["Efficient frontier", "Portfolio", "Tangency port.", "Stocks", "Autoupdate", "off"], fontsize=16, loc="upper left")
plt.xlim([0, 0.05]) # Set x-axis limits as in original
plt.ylim([np.mean(Rf), 0.009]) # Set y-axis limits, starting from the risk-free rate

# Add dashed lines to highlight the Tangency Portfolio's position
plt.axhline(mu_t.flatten(), linestyle='--', color='k', linewidth=1.5) # Ensure 1D for plotting
plt.axvline(sig_t.flatten(), linestyle='--', color='k', linewidth=1.5) # Ensure 1D for plotting

# Save the plot as an EPS file
plt.savefig("MV_rf.eps", format="eps")
# plt.show() # Uncomment to display the plot

# --- Optimal weights Calculation ---
# Define a range of target expected returns for which to calculate optimal weights
mup_target = np.linspace(0.001, 0.05, 10)
# Calculate optimal weights for portfolios on the CML (combination of risk-free and tangency)
# Formula: w_p = inv(S) * (z - rf*1) * (mu_p - rf) / H
# Using proper matrix operations
wp = V1 @ (z - np.ones((N, 1)) * np.mean(Rf)) @ ((mup_target - np.mean(Rf)) / H)

# Calculate the weights for the Tangency Portfolio (risky assets only)
# Formula: w_T = inv(S) * (z - rf*1) / (1' * inv(S) * (z - rf*1))
# Using proper matrix operations
wT = V1 @ (z - np.ones((N, 1)) * np.mean(Rf)) / (np.ones((1, N)) @ V1 @ (z - np.ones((N, 1)) * np.mean(Rf)))

# Calculate the weight allocated to the risk-free asset for each portfolio (Tangency and CML)
# rf_weight = 1 - sum of risky asset weights in each portfolio
# Sum of risky weights for Tangency Portfolio (wT)
sum_wT = np.sum(wT)
# Sum of risky weights for CML portfolios (wp) along axis 0
sum_wp = np.sum(wp, axis=0)
# Combine into a single row vector: [1 - sum(wT), 1 - sum(wp[:, 0]), ..., 1 - sum(wp[:, 9])]
rf_weight_row = np.hstack([1 - sum_wT, 1 - sum_wp])
# Reshape to a row vector of shape (1, 1 + len(mup_target))
rf_weight_row = rf_weight_row.reshape(1, -1)

# --- Display and Save Weights ---
# Create the full weight data matrix by vertically stacking risky weights and the risk-free weight row
# Risky weights matrix: horizontally stack wT (N, 1) and wp (N, len(mup_target)) -> shape (N, 1 + len(mup_target))
risky_weights_matrix = np.hstack([wT, wp])
# Vertically stack the risky weights matrix and the risk-free weight row
weight_data = np.vstack([risky_weights_matrix, rf_weight_row])

# Round the weight data to 3 decimal places as in the original script
weight_data = np.round(weight_data, 3)

# Define column names (Tangency Portfolio + target return percentages)
columns = ["wT"] + [f"{round(mup_val * 100, 2)}%" for mup_val in mup_target]
# Define row names (asset names + Rf)
index = Ret.columns[1:].tolist() + ["Rf"]

# Create a pandas DataFrame to store the calculated weights
Weights = pd.DataFrame(weight_data, columns=columns, index=index)

print("Optimal Weights (Tangency and CML):")
print(Weights)

# Write the weights table to an Excel file
# Save to "Weights_MVrf.xlsx" with index=True and header=True (default behavior)
Weights.to_excel("Weights_MVrf.xlsx", index=True, header=True)


# This Python script contains code examples and material from the book:
# "Essentials of Financial Economics" by Michael Donadelli, Michele Costola, and Ivan Gufler.
# You can find more information and download the book at: link.springer.com/book/9783031861895
